networkx/algorithms/summarization.py

```diff
@@ -210,3 +209,627 @@ def dedensify(G, threshold, prefix=None, copy=True):
+                    if edge_values:
+                        edge_attributes = dict(zip(edge_attributes, edge_values))
+                    else:
+                        edge_attributes = dict()
+                    summary_source_node = node_label_lookup[group_id]
+                    summary_target_node = node_label_lookup[other_group_id]
+                    summary_graph_edge = (summary_source_node, summary_target_node)
+                    output.add_edge(*summary_graph_edge, **edge_attributes)
+
+    if deterministic_labels:
+        # get supernodes
+        supernode_keys = []
+        for key in supernode_lookup:
+            if len(supernode_lookup[key]) > 1:
+                supernode_keys.append(key)
+
+        supernode_keys = sorted(
+            supernode_keys, key=lambda k: list(supernode_lookup[k])[0]
+        )
+        original_supernode_lookup = dict(**supernode_lookup)
+        label_mapping = dict()
+        for index, key in enumerate(supernode_keys):
+            new_key = "%s%s" % (prefix, index)
+            label_mapping[key] = new_key
+            supernode_lookup[new_key] = original_supernode_lookup.pop(key)
+        output = nx.relabel_nodes(output, label_mapping)
+
+    return output, supernode_lookup
+
+
+def snap_build_data_structures(G, node_attributes, edge_attributes):
+    """Returns data structures needed to build the SNAP summary graph
+
+    Creates the data structures and splits the node groups  until all are groups
+    are homogenous in terms of node attribute values and edge types to
+    build the output graph for the SNAP summary graph.
+
+    Parameters
+    ----------
+    G: graph
+        Networkx Graph to be summarized
+    edge_attributes: iterable
+        The edge attributes that can be found in edges in the graph, and
+        are to be recognized in the summarization
+
+    Returns
+    -------
+    tuple of data structure representations: (list, list, dict, dict)
+
+    Notes
+    -----
+    Finds the AR-compatible grouping of graph `G` and returns data structures
+    to efficiently process those groupings.
+
+    An Attribute-edge compatible (AR-compatible) grouping is where
+    all nodes in each group have the same exact node attribute values and
+    have the same exact edge types to nodes in the same exact
+    groups.
+    """
+    if edge_attributes:
+        unique_edge_types = {
+            snap_edge_type_id(G, edge, edge_attributes) for edge in G.edges
+        }
+    else:
+        unique_edge_types = [None]
+    unique_edge_types = list(unique_edge_types)
+    edge_index_lookup = dict(
+        (value, index) for index, value in enumerate(unique_edge_types)
+    )
+
+    group_sets = snap_attribute_compatible_groups(G, node_attributes)
+    group_ids = list(range(len(group_sets)))
+
+    group_lookup = dict()
+    for group_id, group in zip(group_ids, group_sets):
+        for node in group:
+            group_lookup[node] = group_id
+
```

Comment:
Use `enumerate` to finesse the double-list creation + zip idiom.  A dict-comprehension also helps.
```suggestion
group_lookup = {node: gid for gid, group in enumerate(group_sets) for node in group}
```

Reply:
Good idea.  I changed the `supernodes` to be stored in an attribute node called `group`.  I couldn't come up with a better name for it.  To account for potential conflicts with developer attributes, I added a `supernode_attribute` for changing the attribute name.

Reply:
+1 for flexible name of supernode attribute